### PR TITLE
Basics

### PR DESCRIPTION
**Pull request by icy**
This PR is not ready. I will keep updating.

@Nhamlh: Feel free to take a look. Thanks :)

### CODE REVIEW

**Comment by icy** [comment]
@NhamLH: https://github.com/icyfork/vietnamese/blob/f31537075be7941deeeec7ea54409ccc61daf3aa/basics/imports-and-modules.md

**Comment by nhamlh** [comment]
@icy the term `gói` appeared the first time in [here](https://github.com/icyfork/vietnamese/blob/f31537075be7941deeeec7ea54409ccc61daf3aa/basics/imports-and-modules.md#gi%E1%BB%9Bi-h%E1%BA%A1n-khi-n%E1%BA%A1p-m%C3%B4-%C4%91un) but the meaning of it is unclear. Please help clarify it.

File: basics/basic-types.md [hunk 1]
```diff
@@ -1,91 +1,97 @@
...
-in UTF-32 strings.
+Kiểu bắt đầu bởi `u` để kiểu không nhận giá trị âm (không dấu).
+Kiểu `char` thực sự được nhúng trong UTF-8, trong khi `wchar`
+dùng ký tự UTF-16, còn `dchar` là UTF-32.
 
-A conversion between variables of different types is only
-allowed by the compiler if no precision is lost. However,
-a conversion between floating point types
-(e.g `double` to `float`) is allowed.
+Trình biên dịch D chỉ thực hiện chuyển đổi kiểu nếu đảm bảo được sự toàn vẹn
```
Comment (nhamlh) [inline basics/basic-types.md]: I think `cho phép` has a more precise meaning than `thực hiện` in this case.

File: basics/basic-types.md [hunk 2]
```diff
@@ -1,91 +1,97 @@
...
 |---------|--------------------------------------------------
 |`float`  | 32-bit
 |`double` | 64-bit
-|`real`   | >= 64-bit (generally 64-bit, but 80-bit on Intel x86 32-bit)
+|`real`   | >= 64-bit (thường là 64-bit, hoặc 80-bit trên Intel x86 32-bit)
 
-The prefix `u` denotes *unsigned* types. `char` translates to
-UTF-8 characters, `wchar` is used in UTF-16 strings and `dchar`
-in UTF-32 strings.
+Kiểu bắt đầu bởi `u` để kiểu không nhận giá trị âm (không dấu).
```
Comment (nhamlh) [inline basics/basic-types.md]: maybe `u` để biểu thị kiểu :D

File: basics/basic-types.md [hunk 3]
```diff
@@ -1,91 +1,97 @@
...
-In D, indexes usually have the alias type `size_t`, as it is a type that
-is large enough to represent an offset into all addressable memory - this is
-`uint` for 32-bit and `ulong` for 64-bit architectures.
+    int.max # 2147483647
+    int.min # -2147483648
 
-### Assert expression
+Với kiểu số chấm động, `kiểu.min_normal` là giá trị  nhỏ nhất khác `0`
+biểu diễn được trong kiểu, còn `.nan` chỉ giá trị không xác định,
+`.infinity` chỉ vô cùng, `.dig` chỉ số chữ số sau dấu phấy thập phân,
```
Comment (nhamlh) [inline basics/basic-types.md]: Please fix typo

File: basics/type-qualifiers.md [hunk 4]
```diff
@@ -1,66 +1,72 @@
...
-`immutable` objects can thus be safely shared among different threads with no
-synchronization because they never change by definition. This also implies that
-`immutable` objects can be cached perfectly.
+### `hằng`
 
-### `const`
+Biến khai báo với `const` cũng không thể thay đổi giá trị, nhưng điều này
+chỉ áp dụng trong giới hạn của ngữ cảnh hiện tại (_scope_). Ví dụ, một
+biến được khai báo giá trị ở một nơi, và được dùng trong một hàm
+với ràng buộc rằng biến không thể giá trị bên trong hàm đó.
```
Comment (nhamlh) [inline basics/type-qualifiers.md]: Is it 'không thể thay đổi giá trị'?

File: basics/basic-types.md (hunk 1 above)
Comment (icy) [reply basics/basic-types.md]: Fixed in translation.

**Comment by icy** [comment]
Hi @wilzbach , this PR has been merged however the home page https://tour.dlang.org/tour/vi is not updated. Am I missing something? Thx a lot.

**Comment by wilzbach** [comment]
Yep, the CI is currently broken because of a dependency conflict with mir :/

https://travis-ci.org/dlang-tour/core

I will be busy for today, but I will try to have a look into this tomorrow. Sorry.

**Comment by ZombineDev** [comment]
@icy I'll try to have have a look today. @wilzbach I have been working on some changes relating to Docker (e.g. multi-stage building), which will allow building locally on Linux distros that don't ship with static libraries, like my Arch Linux. But I also need to first debug the mir issue, before I can post my PR.

**Comment by icy** [comment]
thanks a lot @wilzbach  and @ZombineDev  for your updates. Pls take your time :)

**Comment by ZombineDev** [comment]
You're welcome @icy, glad to help :)

**Pull request merged**